drafts/LET-DRAFT.md

```diff
@@ -0,0 +1,572 @@
+* * *
+
+## Background
+
+The original discussion around this issue was inspired by the Github issue [#653](https://github.com/partiql/partiql-lang-kotlin/issues/653) and [#549](https://github.com/partiql/partiql-lang-kotlin/issues/549), which can be summarized by the following table.
+
+|Query	|CompilerPipeline Actual	|PlannerPipeline Actual	|
+|---	|---	|---	|
+|`SELECT f.* FROM << {'a':1} >> AS f LET 23 AS f`	|`<< { '_1' : 23 } >>`	|`<< { 'a' : 1 } >>`	|
+|`SELECT * FROM << {'a':1} >> AS f LET 23 AS f`	|`<< { '_1' : 23 } >>`	|`<< { 'a' : 1 } >>`	|
+
+To deicide the expected behavior, we need to figure out the definition for `LET` and `SELECT *`.
+* * *
+
+## Overview
+
+In this doc, I first try to define the `LET` Statement based on the existing spec, then discussed two alternative approaches that may require small addition to the Spec. Next, I discussed the possibility of extending the support of  `LET` to other clauses and proposed how `SELECT *` could interact with variables defined by `LET`. Finally, I did a cross comparison on results of the above queries based on different design.
+* * *
+
+## **Terminology**
+
+In this document, I tried to stay coherent with the terminology defined in our [spec](https://partiql.org/assets/PartiQL-Specification.pdf).
+
+1. Binding tuple: `bi = ⟨ x1 : v1, ..... xn : vn⟩` where each `xi` is a **bind name** to the PartiQL value `vi`
+2. The bag of binding tuples: `<<b1,......bk>>` denoted B<sup>in/out</sup><sub>Clause</sub>
+3. `⍴0,⍴ ⊢ q -> v` denotes that the PartiQL query `q` evaluates to the value `v` when evaluated within the database environment `⍴0` and variable environment, i.e. when every variable of `q` is instantiated by its binding in `⍴` and each database name is instantiated to its value in `⍴0`.
+4. Variables environment: denoted by `⍴` created by the defined query variables.
+
+* * *
+
+## Definition of `LET`
+
+For now, consider the scope is limit to `FROM-LET`, i.e., `FROM` clause followed by `LET` .
+
+### **Use case of LET:**
+
+LET should be used as a way to reduce redundancy and complexity of the query. It is sometimes useful to store the result of a (sub) expression / (sub) query to use in a subsequent clause. `LET` should allow such “variables storage”, by creating new variable bindings based on the collection of bindings fed in through it as part of the `SELECT`-`FROM`-`WHERE`.
+
+Example:
+
+```
+select * from s3object
+where (((((not (lower("payload") like '%gurr%' escape '\'))
+and (not (lower("payload") like '%[stream_shadow=true]%' escape '\'))
+and (((not (lower("payload") like '%gurr%' escape '\'))
+and lower("_sourcecategory") = 'stream' and lower("payload") like '%handling%' escape '\'
+and lower("payload") like '%failure%' escape '\' and lower("payload") like '%error%' escape '\'
+......
+```
+
+With let, we can rewrite the query to something similar to:
+
+```SQL
+SELECT *
+from s3object as s3
+let lower("payload") as x
+where not x like ...
+
+* * *
```

Comment:
I think GitHub suggestion has some difficulty w/ rendering 3 backticks. We need to close the block here.